mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java

```diff
@@ -549,6 +550,105 @@ public void setFlagsShouldUpdateFlagsWhenMessageIsInTwoMailboxes() throws Except
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+    }
+
+    @ContractTest
+    public void setFlagsShouldWorkWhenCalledOnDuplicatedMailbox() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+        message2.setUid(mapperProvider.generateMessageUid());
+        message2.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message2);
+        message3.setUid(mapperProvider.generateMessageUid());
+        message3.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message3);
+        message4.setUid(mapperProvider.generateMessageUid());
+        message4.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message4);
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(messageId, ImmutableList.of(message1.getMailboxId(), message1.getMailboxId()), new Flags(Flag.ANSWERED), FlagsUpdateMode.ADD);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+    }
+
+    @ContractTest
+    public void setFlagsShouldWorkWithConcurrencyWithAdd() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+
+        int threadCount = 2;
+        int updateCount = 10;
+        new ConcurrentTestRunner(threadCount, updateCount) {
+            @Override
+            protected void performOperation(int threadNumber, int step) throws Exception {
+                sut.setFlags(message1.getMessageId(),
+                    ImmutableList.of(message1.getMailboxId()),
+                    new Flags("custom-" + threadNumber + "-" + step),
+                    FlagsUpdateMode.ADD);
+            }
+        }.run();
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(message1.getMessageId()), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).createFlags().getUserFlags()).hasSize(threadCount * updateCount);
+    }
+
+    @ContractTest
+    public void setFlagsShouldWorkWithConcurrencyWithRemove() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        message1.setModSeq(mapperProvider.generateModSeq(benwaInboxMailbox));
+        sut.save(message1);
+
+        final int threadCount = 4;
+        final int updateCount = 20;
+        new ConcurrentTestRunner(threadCount, updateCount) {
+            @Override
+            protected void performOperation(int threadNumber, int step) throws Exception {
+                if (step  < updateCount / 2) {
```

Comment:
I don't get why you don't chain the two calls in this method instead of this if/else logic